libcudacxx/.upstream-tests/test/cuda/barrier_arrive_tx.pass.cpp

```diff
@@ -0,0 +1,124 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: pre-sm-90
+
+#include <cuda/barrier>
+
+#include "cuda_space_selector.h"
+#include "test_macros.h"
+#include <cstdint>              // uint and friends
+#include <utility>              // std::move
+
+// Suppress warning about barrier in shared memory
+TEST_NV_DIAG_SUPPRESS(static_var_with_dynamic_init)
+
+using barrier = cuda::barrier<cuda::thread_scope_block>;
+
+__device__ bool run_arrive_tx_test(barrier &bar) {
+  if (threadIdx.x == 0) {
+      init(&bar, blockDim.x);
+  }
+  __syncthreads();
+
+  auto token = bar.arrive_tx(1, 1);
+
+  // Manually increase the transaction count of the barrier by blockDim.x.
+  // This emulates a cp.async.bulk instruction or equivalently, a memcpy_async call.
+  if (threadIdx.x == 0) {
+      asm volatile(
+          "mbarrier.complete_tx.relaxed.cta.shared::cta.b64 [%0], %1;"
+          :
+          : "r"((uint32_t) __cvta_generic_to_shared(cuda::device::barrier_native_handle(bar)))
+          , "r"(blockDim.x)
+          : "memory");
+  }
+  bar.wait(std::move(token));
+
+  return true;
+}
+
+
+
+#ifdef __CUDACC_RTC__
```

Comment:
We have started moving to more expressive macros for those conditional compilations
 
```suggestion
#ifdef TEST_COMPILER_NVRTC
```